internal/authentication/azure/auth.go

```diff
@@ -105,94 +105,158 @@ func (s EnvironmentSettings) GetAzureEnvironment() (*cloud.Configuration, error)
-		} else {
-			errs = append(errs, credErr)
+func (s EnvironmentSettings) addCLIProvider(timeout time.Duration, creds *[]azcore.TokenCredential, errs *[]error) {
+	cred, credErr := azidentity.NewAzureCLICredential(nil)
+	if credErr == nil {
+		*creds = append(*creds, &timeoutWrapper{cred: cred, authmethod: "Azure CLI", timeout: 30 * time.Second})
+	} else {
+		*errs = append(*errs, credErr)
+	}
+}
+
+func (s EnvironmentSettings) addProviderByAuthMethodName(authMethod string, creds *[]azcore.TokenCredential, errs *[]error) {
+	switch authMethod {
+	case "clientcredentials", "creds":
+		s.addClientCredentialsProvider(creds, errs)
+	case "clientcertificate", "cert":
+		s.addClientCertificateProvider(creds, errs)
+	case "workloadidentity", "wi":
+		s.addWorkloadIdentityProvider(creds, errs)
+	case "managedidentity", "mi":
+		s.addManagedIdentityProvider(1*time.Second, creds, errs)
+	case "commandlineinterface", "cli":
+		s.addCLIProvider(30*time.Second, creds, errs)
+	}
+}
+
+func getAzureAuthMethods() []string {
+	return []string{"clientcredentials", "creds", "clientcertificate", "cert", "workloadidentity", "wi", "managedidentity", "mi", "commandlineinterface", "cli", "none"}
+}
+
+// GetTokenCredential returns an azcore.TokenCredential retrieved from the order specified via
+// the azureAuthMethods component metadata property which denotes a comma-separated list of auth methods to try in order.
+// The possible values contained are (case-insensitive):
+// ServicePrincipal, Certificate, WorkloadIdentity, ManagedIdentity, CLI
+// The string "None" can be used to disable Azure authentication.
+//
+// If the azureAuthMethods property is not present, the following order is used (which with the exception of step 5
+// matches the DefaultAzureCredential order):
+// 1. Client credentials
+// 2. Client certificate
+// 3. Workload identity
+// 4. MSI (we use a timeout of 1 second when no compatible managed identity implementation is available)
+// 5. Azure CLI
+func (s EnvironmentSettings) GetTokenCredential() (azcore.TokenCredential, error) {
+	// Create a chain
+	var creds []azcore.TokenCredential
+	errs := make([]error, 0, 3)
+
+	authMethods, ok := s.GetEnvironment("AzureAuthMethods")
+	if !ok || strings.TrimSpace(authMethods) == "" {
+		// 1. Client credentials
+		s.addClientCredentialsProvider(&creds, &errs)
+
+		// 2. Client certificate
+		s.addClientCertificateProvider(&creds, &errs)
+
+		// 3. Workload identity
+		s.addWorkloadIdentityProvider(&creds, &errs)
+
+		// 4. MSI with timeout of 1 second (same as DefaultAzureCredential)
+		s.addManagedIdentityProvider(1*time.Second, &creds, &errs)
+
+		// 5. AzureCLICredential
+		s.addCLIProvider(30*time.Second, &creds, &errs)
+	} else {
+		authMethodIdentifiers := getAzureAuthMethods()
+		authMethods := strings.Split(strings.ToLower(strings.TrimSpace(authMethods)), ",")
+		for _, authMethod := range authMethods {
```

Comment:
QQ: What if one passes both "ManagedIdentity" and "MI"? Wouldn't that cauise the auth method to be added twice?

Reply:
That is correct - and that's totally ok! You might want it to be tried twice for some reason. You have full control :)

You can have write `mi,mi,mi,mi,mi` it's all up to you :smile